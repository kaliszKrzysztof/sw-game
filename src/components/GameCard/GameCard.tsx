import React from 'react';
import Card from '@material-ui/core/Card';
import CardHeader from '@material-ui/core/CardHeader';
import CardContent from '@material-ui/core/CardContent';
import Typography from '@material-ui/core/Typography';
import Collapse from '@material-ui/core/Collapse';
import { Attribute } from 'types/game';

export type GameCardStatus = 'win' | 'lose' | 'draw';

interface GameCardProps {
  title: string;
  attributes: Attribute[];
  expanded: boolean;
  status?: GameCardStatus;
}

const getStyles = (status: GameCardStatus): React.CSSProperties => {
  switch (status) {
    case 'win': {
      return {
        borderColor: '#149414',
        color: '#149414',
      };
    }
    case 'lose': {
      return {
        borderColor: '#F44336',
        color: '#F44336',
      };
    }
    case 'draw': {
      return {
        borderColor: '#FCCF14',
        color: '#FCCF14',
      };
    }
    default: {
      return {};
    }
  }
};

const GameCard: React.FC<GameCardProps> = ({
  title,
  attributes,
  expanded,
  status,
}) => {
  return (
    <Card variant="outlined" style={getStyles(status)} data-testid="game-card">
      <CardHeader
        title={title}
        titleTypographyProps={{ align: 'center', color: 'inherit' }}
      />
      {attributes.length > 0 && (
        <Collapse in={expanded} timeout="auto" unmountOnExit>
          <CardContent data-testid="attributes">
            {attributes.map(({ name, value }, index) => (
              <Typography
                data-testid="attribute"
                color="inherit"
                key={name}
                align="center"
                paragraph={index < attributes.length - 1}
              >{`${name}: ${value}`}</Typography>
            ))}
          </CardContent>
        </Collapse>
      )}
    </Card>
  );
};

export default GameCard;
