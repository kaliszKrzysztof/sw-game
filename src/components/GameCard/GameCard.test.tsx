import React from 'react';
import { render } from '@testing-library/react';
import GameCard from './GameCard';

const title = 'test';
const attributes = [
  {
    name: 'attr1',
    value: 'attr1',
  },
  {
    name: 'attr2',
    value: 'attr2',
  },
  {
    name: 'attr3',
    value: 'attr3',
  },
  {
    name: 'attr4',
    value: 'attr4',
  },
];

describe('<GameCard />', () => {
  it('should render', () => {
    render(<GameCard title={title} attributes={attributes} expanded={false} />);
  });

  it('should match snapshot', () => {
    const { container } = render(
      <GameCard title={title} attributes={attributes} expanded={false} />,
    );
    expect(container).toMatchSnapshot();
  });

  it('renders all passed attributes if expanded', async () => {
    const { queryByTestId, getAllByTestId, rerender } = render(
      <GameCard title={title} attributes={attributes} expanded={false} />,
    );
    expect(queryByTestId('attributes')).not.toBeInTheDocument();

    rerender(<GameCard title={title} attributes={attributes} expanded />);
    const renderedAttributes = getAllByTestId('attribute').map(
      (element) => element.textContent,
    );
    const allAttributes = attributes.map(
      ({ name, value }) => `${name}: ${value}`,
    );
    expect(renderedAttributes).toEqual(allAttributes);
  });

  it('renders status related styles', async () => {
    const { getByTestId, rerender } = render(
      <GameCard title={title} attributes={attributes} expanded={false} />,
    );
    expect(getByTestId('game-card')).not.toHaveAttribute('style');
    rerender(
      <GameCard
        title={title}
        attributes={attributes}
        expanded={false}
        status="win"
      />,
    );
    expect(getByTestId('game-card')).toHaveAttribute(
      'style',
      'border-color: #149414; color: rgb(20, 148, 20);',
    );
    rerender(
      <GameCard
        title={title}
        attributes={attributes}
        expanded={false}
        status="lose"
      />,
    );
    expect(getByTestId('game-card')).toHaveAttribute(
      'style',
      'border-color: #f44336; color: rgb(244, 67, 54);',
    );
    rerender(
      <GameCard
        title={title}
        attributes={attributes}
        expanded={false}
        status="draw"
      />,
    );
    expect(getByTestId('game-card')).toHaveAttribute(
      'style',
      'border-color: #fccf14; color: rgb(252, 207, 20);',
    );
  });
});
