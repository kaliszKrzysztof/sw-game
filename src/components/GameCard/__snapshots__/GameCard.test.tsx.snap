// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GameCard /> should match snapshot 1`] = `
<div>
  <div
    class="MuiPaper-root MuiCard-root MuiPaper-outlined MuiPaper-rounded"
    data-testid="game-card"
  >
    <div
      class="MuiCardHeader-root"
    >
      <div
        class="MuiCardHeader-content"
      >
        <span
          class="MuiTypography-root MuiCardHeader-title MuiTypography-h5 MuiTypography-colorInherit MuiTypography-alignCenter MuiTypography-displayBlock"
        >
          test
        </span>
      </div>
    </div>
  </div>
</div>
`;
